libbitstream/bitstream.c

```diff
@@ -0,0 +1,313 @@
+// Copyright(c) 2018-2019, Intel Corporation
+//
+// Redistribution  and  use  in source  and  binary  forms,  with  or  without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of  source code  must retain the  above copyright notice,
+//   this list of conditions and the following disclaimer.
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+// * Neither the name  of Intel Corporation  nor the names of its contributors
+//   may be used to  endorse or promote  products derived  from this  software
+//   without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
+// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif // HAVE_CONFIG_H
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <uuid/uuid.h>
+#include <json-c/json.h>
+#include "bitstream.h"
+
+#include "safe_string/safe_string.h"
+#include <opae/log.h>
+#include <opae/properties.h>
+#include <opae/sysobject.h>
+
+#define METADATA_GUID     "58656F6E-4650-4741-B747-425376303031"
+#define METADATA_GUID_LEN 16
+#define GBS_AFU_IMAGE     "afu-image"
+#define BBS_INTERFACE_ID  "interface-uuid"
+
+/*
+ * Check for bitstream header and fill out bistream_info fields
+ */
+#define MAGIC       0x1d1f8680
+#define MAGIC_SIZE  4
+#define HEADER_SIZE 20
+STATIC int parse_metadata(opae_bitstream_info *info)
+{
+	unsigned i;
+
+	if (!info)
+		return -EINVAL;
+
+	if (info->data_len < HEADER_SIZE) {
+		OPAE_ERR("File too small to be GBS.");
+		return -1;
+	}
+
+	if (((uint32_t *)info->data)[0] != MAGIC) {
+		OPAE_ERR("No valid GBS header.");
+		return -1;
+	}
+
+	/* reverse byte order when reading GBS */
+	for (i = 0; i < sizeof(info->interface_id); i++)
```

Comment:
Is this 128-bit? Is this really reversed like this or should it be bswap64 on the two components?

Reply:
Will note as a TODO for @anandaravuri to evaluate.